File: src/main/scala/com/databricks/labs/mosaic/datasource/multiread/MosaicDataFrameReader.scala
package com.databricks.labs.mosaic.datasource.multiread

import org.apache.spark.sql.{DataFrame, SparkSession}

/**
  * A Mosaic DataFrame Reader that provides a unified interface to read multiple
  * formats. Currently supports:
  *   - OGR Multi Read
  *   - Raster to Grid
  * @param sparkSession
  *   The Spark Session to use for reading. This is required to create the
  *   DataFrame.
  */
class MosaicDataFrameReader(sparkSession: SparkSession) {

    /** Extra options to pass to the underlying reader. */
    protected var extraOptions: Map[String, String] = Map.empty[String, String]

    /** The format to use for reading. */
    protected var format: String = ""

    /**
      * Set a single option.
      * @param key
      *   The key for the option.
      * @param value
      *   The value for the option.
      * @return
      *   This MosaicDataFrameReader.
      */
    def option(key: String, value: String): MosaicDataFrameReader = {
        extraOptions += (key -> value)
        this
    }

    /**
      * Set a map of options. This will overwrite any existing options.
      * @param map
      *   The map of options to set.
      * @return
      *   This MosaicDataFrameReader.
      */
    def options(map: Map[String, String]): MosaicDataFrameReader = {
        extraOptions = map
        this
    }

    /**
      * Set the format to use for reading.
      * @param source
      *   The format to use.
      * @return
      *   This MosaicDataFrameReader.
      */
    def format(source: String): MosaicDataFrameReader = {
        this.format = source
        this
    }

    /**
      * Load a single path.
      * @param path
      *   The path to load.
      * @return
      *   The DataFrame.
      */
    def load(path: String): DataFrame = {
        format match {
            case "multi_read_ogr" => new OGRMultiReadDataFrameReader(sparkSession)
                    .options(extraOptions)
                    .asInstanceOf[OGRMultiReadDataFrameReader]
                    .load(path)
            case "raster_to_grid" => new RasterAsGridReader(sparkSession)
                    .options(extraOptions)
                    .asInstanceOf[RasterAsGridReader]
                    .load(path)
            case _                => throw new Error(s"Unsupported format: $format")
        }
    }

    /**
      * Load multiple paths.
      * @param paths
      *   The paths to load.
      * @return
      *   The DataFrame.
      */
    def load(paths: String*): DataFrame = {
        format match {
            case "multi_read_ogr" => new OGRMultiReadDataFrameReader(sparkSession)
                    .options(extraOptions)
                    .asInstanceOf[OGRMultiReadDataFrameReader]
                    .load(paths: _*)
            case "raster_to_grid" => new RasterAsGridReader(sparkSession)
                    .options(extraOptions)
                    .asInstanceOf[RasterAsGridReader]
                    .load(paths: _*)
            case _                => throw new Error(s"Unsupported format: $format")
        }
    }

}

File: python/mosaic/readers/mosaic_data_frame_reader.py
from pyspark.sql import SparkSession, DataFrame, SQLContext


class MosaicDataFrameReader:
    """
    MosaicDataFrameReader is a wrapper around the MosaicDataFrameReader that allows for offset reading
    of OGR data sources. This provides a means to read data sources that are too large to fit into memory
    in a single read. This is accomplished by reading the data source in chunks and using the generators.
    """

    def __init__(self):
        """
        Initialize the MosaicDataFrameReader.
        """
        self.spark = SparkSession.builder.getOrCreate()
        self.reader = getattr(
            self.spark._jvm.com.databricks.labs.mosaic.datasource.multiread, "MosaicDataFrameReader"
        )(self.spark._jsparkSession)

    def format(self, format):
        """
        Set the format of the data source to read.
        :param format: Format of the data source to read.
        :return: MosaicDataFrameReader with the param set to provided value.
        """
        self.reader.format(format)
        return self

    def option(self, key, value):
        """
        Set the option for the data source to read.
        :param key: Option key.
        :param value: Option value.
        :return: MosaicDataFrameReader with the param set to provided value.
        """
        self.reader.option(key, value)
        return self

    def load(self, path):
        """
        Load the data source as a MosaicDataFrame.
        :param path: Path to the data source.
        :return: MosaicDataFrame.
        """
        df = self.reader.load(path)
        return DataFrame(df, SQLContext(self.spark.sparkContext))

    def load(self, *paths):
        """
        Load the data source as a MosaicDataFrame.
        :param paths: Paths to the data source.
        :return: MosaicDataFrame.
        """
        df = self.reader.load(*paths)
        return DataFrame(df, SQLContext(self.spark.sparkContext))
